### ISSUE TITLE
Platform Artifact Support

### ISSUE THREAD

**Issue by sghill** [nebula-plugins/gradle-nebula-integration#14]
project: https://github.com/nebula-plugins/gradle-nebula-integration/tree/master/excludes-in-bom

Excludes within BOM
===================

A dependency that is excluded within a bom ends up brought in anyway.

BOM
---

```
<dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>org.springframework</groupId>
            <artifactId>spring-beans</artifactId>
            <version>5.0.7.RELEASE</version>
            
            <exclusions>
                <exclusion>
                    <groupId>org.yaml</groupId>
                    <artifactId>snakeyaml</artifactId>
                </exclusion>
            </exclusions>
            
        </dependency>
    </dependencies>
</dependencyManagement>
```

build.gradle
------------

```
dependencies {
  implementation 'nebulatest:nebulabom:1.0.0'       // recommends a version of spring-beans excluding snakeyaml
  implementation 'org.springframework:spring-beans' // expecting snakeyaml to be excluded
  implementation 'org.yaml:snakeyaml:1.19'          // snakeyaml loses to spring-bean's snakeyaml
}
```

Dependencies Resolved
---------------------

```
$ ./gradlew -q dependencies --configuration compileClasspath

------------------------------------------------------------
Root project
------------------------------------------------------------

compileClasspath - Compile classpath for source set 'main'.
+--- nebulatest:nebulabom:1.0.0
|    \--- org.springframework:spring-beans:5.0.7.RELEASE
|         +--- org.springframework:spring-core:5.0.7.RELEASE
|         |    \--- org.springframework:spring-jcl:5.0.7.RELEASE
|         \--- org.yaml:snakeyaml:1.20                           <-- expected to be excluded
+--- org.springframework:spring-beans -> 5.0.7.RELEASE (*)
\--- org.yaml:snakeyaml:1.19 -> 1.20

(*) - dependencies omitted (listed previously)
```

Ideally I'd like to be able to craft a BOM that can exclude all the known edges that bring in a problematic dependency. Is it expected this would not work today?

I suspect this may be because `dependencyManagement` maps to [DependencyConstraint](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/DependencyConstraint.html), which doesn't have any exclude abilities, but wanted to check.

**Comment by bigdaz** [comment]
There are a couple of reasons why an `exclusion` defined in a `dependencyManagement` block doesn't get applied, even when `IMPROVED_POM_SUPPORT` is enabled. As you suspect, it's related to the fact that Gradle creates `constraints` for each entry in the `dependencyManagement` block for a "BOM" (defined as `packaging == 'pom' and no artifact).

However, on inspection I can see that the exclusions are intact for these dependency constraints, due to the shared internal implementation of hard dependencies and dependency constraints. So even though you can't define an exclude for a dependency constraint in the DSL, we do define these excludes them when parsing a BOM, and retain them when processing the dependency graph.

There are 2 reasons these don't take effect like you expect:
1) Dependency constraints are defined as non-transitive, and we currently ignore dependency exclusions defined for any non-transitive dependency.
2) Dependency constraints effectively create an _additional path_ to the dependency node in the graph, and Gradle only excludes a transitive dependency if it's excluded from _all incoming paths_. See https://github.com/gradle/gradle/issues/1473#issuecomment-342938454.

I've experimented with and spiked 2 possible solutions to this issue. (Note that for the purpose of the example, please ignore the fact that `snakeyaml` is an _optional_ dependency of `spring-beans`).

#### 1. Special treatment for any `exclusion` defined in a `dependencyManagement` entry

With this behaviour, any `exclusion` defined in a `dependencyManagement` dependency (or any constraint) would apply globally to the target node. This is the equivalent of applying the exclusion to all incoming edges for the node. So if you excluded `org.yaml:snakeyaml` from `spring-beans`, then `snakeyaml` would _never_ be brought in transitively by `spring-beans`, no matter how `spring-beans` was added to the graph. 

This implementation makes the BOM dependency somewhat viral, since as soon as the BOM dependency is in the graph, it will impact all uses of `spring-beans`.  We've specifically avoided this sort of viral behaviour for version management with BOM import, since we don't want for transitive consumers of a BOM to be unable to escape the defined versions without resorting to `force`. For excludes it doesn't feel quite as "nuclear", since a consumer can explicitly add a dependency on `snakeyaml` (as shown in your example above), and it will be resolved with the version specified.

One downside to this approach is that it does not apply to `exclusions` defined in a regular dependency in a BOM: so if your BOM added a hard dependency on `spring-beans` with an exclusion on `snakeyaml` and a build script added another hard dependency on `spring-beans` without, then the latter dependency would drag in the version of `snakeyaml` defined in the `spring-beans` POM.

#### 2. Special treatment for dependencies with empty versions

Quite often, a BOM dependency is combined with a bunch of dependency declarations that omit versions. An alternative solution to the `dependencyManagement - exclusion` conundrum is to consider than any dependency without a version declared is effectively non-transitive. As such, it would not provide an additional path in the graph, and would not override the dependency exclusions provided by a constraint.

So in your example above, there would be 2 incoming paths to `spring-beans` (one from the BOM and one in the build script), but only the one from the BOM would be considered for the purposes of exclusions. This would make the exclusion work as expected in your example.

There are some downsides to this approach, not the least of which is that it can lead to some surprising behaviour:
- The _effective path_ to `spring-beans` and it's transitive dependencies is via the BOM, not via the directly dependency (declared without a version)
- As soon as a user explicitly adds `implementation 'org.springframework:spring-beans:5.0.7-RELEASE'` to their build script, suddenly the exclusion defined in the BOM would no longer apply.

-----

Note that these 2 options are not exclusive: we could implement special handling for both dependencies with empty versions, and for exclusions on dependency constraints. I think that option #1 is more useful and less surprising at this stage.

**Comment by bigdaz** [comment]
@wilkinsona If you're around it would be good to have your input into this. I'll create a linked issue on https://github.com/gradle/gradle/issues so it's not hidden over here on the nebula issues board.

**Comment by rpalcolea** [comment]
This is fixed in `5.0-20181007235844+0000`

```
compileClasspath - Compile classpath for source set 'main'.
+--- nebulatest:nebulabom:1.0.0
+--- org.springframework:spring-beans:5.1.0.RELEASE
|    \--- org.springframework:spring-core:5.1.0.RELEASE
|         \--- org.springframework:spring-jcl:5.1.0.RELEASE
\--- org.yaml:snakeyaml:1.19

```

vs Gradle 4.10.2

```
------------------------------------------------------------

compileClasspath - Compile classpath for source set 'main'.
+--- nebulatest:nebulabom:1.0.0
|    \--- org.springframework:spring-beans:5.0.7.RELEASE -> 5.1.0.RELEASE
|         +--- org.springframework:spring-core:5.1.0.RELEASE
|         |    \--- org.springframework:spring-jcl:5.1.0.RELEASE
|         \--- org.yaml:snakeyaml:1.23
+--- org.springframework:spring-beans:5.1.0.RELEASE (*)
\--- org.yaml:snakeyaml:1.19 -> 1.23
```

Closing now.

Thanks @bigdaz @ljacomet
